### ISSUE TITLE
Add scripting interface for adding custom SEXPs

### ISSUE THREAD

**Issue by asarium** [scp-fs2open/fs2open.github.com#1320]
We already have the capability of executing scripts from SEXPs but it would be even better to allow modders to add actual SEXPs which could be used just like any other engine SEXP. To reduce the amount of work required, the feature could be limited so that new SEXPs could only be added at a specific time when the game is initializing. Since I would like to display the custom SEXPs in FRED as well, that seems like the best choice since. 

I don't have much experience with the inner workings of the SEXP system so I'll need some guidance from more experience SEXP coders to point me into the right direction.

I'll use this issue to document my progress while implementing this feature but it will probably take some time since the SEXP system is based around the idea that the operator list is static.

**Comment by AxemP** [comment]
I was actually going to suggest something similar... I think. (It sounds sort of related anyway)

Basically my idea would be in a part of a scripting table, you can define a Lua to SEXP interface. Something that would sort of look like this...
https://ibin.co/3EE3QdQAG8ss.png
Description would be help text, and the fields after would correspond to the different types within the sexp system. Number, string, boolean, ships, wings, backgrounds etc etc etc.

So then with a new script-eval-define sexp that seeks and "compiles" (or enumerates) the defined entries in a list, and lets a FREDder choose assignments to the parameters.
https://ibin.co/3EE40EB6iTpS.png
https://ibin.co/3EE4B9gvINbu.png

So when the script-eval-define gets called in mission, the system calls the function with the arguments in the order that they were defined in that initial declaration.

**Comment by X3N0-Life-Form** [comment]
The idea of custom SEXPs has been tossed around for a while now, and I think pretty much everybody agreed that it would be a great enhancement. People get scared off whenever one mentions scripts, regardless of how easy or difficult to use said scripts might be.

In any case, if you need testers, or even just bounce ideas off modders, we'd be happy to help.

**Comment by asarium** [comment]
Thanks for your suggestions. I was going to approach this from the SEXP side first but discussing the mod interface can't hurt.

@AxemP My initial idea was to add a new lua API with which a script could register a lua function as a SEXP. That would have the advantage that the lua function wouldn't have to be defined in the global namespace which would reduce the risk of conflicts with other mods. Here is an example of what I would propose:
```lua
mn.addSEXP("script-sexp", {
    -- this are the possible parameters and their documentation. Not shown here to keep this short
}, function(target_ship, mode_num)
    ba.error("This doesn't work!")
end)
```

This SEXP would appear just like any other SEXP (maybe in a separate category) and wouldn't require any special SEXP to execute.

**Comment by AxemP** [comment]
Sweet, that sounds cool too. Let me know when you need stuff tested. I'll let you code geniuses take care of the technical stuff.

**Comment by asarium** [comment]
Technical update: I vectorized the `Operators` and `Sexp_help` arrays so elements can be added to that at runtime. As far as I can tell, that should cover parsing new SEXPs since that only uses `get_operator_index` which will automatically pick up elements added to the `Operators` vector.

I am now starting to investigate how to integrate dynamically added SEXPs into the existing functions. My current plan is to use the `default` case in the various SEXP operator related functions to check if the current operator index belongs to a dynamically added operator and if that's the case call the appropriate function for that SEXP. I don't know much about the SEXP system at this point so any comments about how this system could be improved would be greatly appreciated.

**Comment by asarium** [comment]
Another update: Lua SEXPs can now be registered and the SEXP system can use them properly, including execution of the registered Lua function. The current code is here: https://github.com/asarium/fs2open.github.com/tree/feature/scriptedSEXPs

However, that only covers the FSO side of the feature. Now I need to fix the things I have broken in FRED and enable it to use the dynamic SEXPs properly. For that I will need to add Lua support to FRED. Currently FRED does not initialize the script system or run any code which means that most mods out there are are not designed with FRED in mind because it never was an issue before. In order to not break those mods I will introduce a `game_settings.tbl` option for enabling Lua in FRED which will require the modder to check that their scripts work properly when run inside FRED (or disable them if FRED is enabled, I'll add a lua API function for checking if that's the case).

**Comment by asarium** [comment]
FRED now also supports dynamic SEXPs!
![luasexps](https://cloud.githubusercontent.com/assets/1013983/24660859/0194ec84-1951-11e7-9c71-3bd7e8445fa4.png)

**Comment by asarium** [comment]
I posted an RFC for deciding how the feature should be integrated into the modding workflow: http://www.hard-light.net/forums/index.php?topic=93399.msg1845849#msg1845849

**Comment by asarium** [comment]
I updated the PR to use a table based solution and I think the changes should be ready for some extended testing now. I have compiled [some builds](https://bintray.com/scp-fs2open/FSO/download_file?file_path=luaSEXP.7z) (64-bit SSE2) and prepared a [test mod](https://bintray.com/scp-fs2open/FSO/download_file?file_path=LuaSEXPs.7z) to demonstrate the usage of the new feature.

I can't post this to the HLP thread since the server are apparently down but I'll add these files to the thread once it comes back.
